model_utilities/train/finetuner.py

```diff
@@ -43,52 +44,83 @@ def __plot_losses(self, training_losses: list[float], validation_losses: list[fl
         self.__log_train_config(config)
 
+        # Prepare variables
         train_conf = config["TRAINING"]
+        batch_size = int(train_conf["batch_size"])
+        max_length = int(train_conf["max_length"])
+        num_epochs = int(train_conf["num_epochs"])
+        shuffle_seed = int(train_conf["shuffle_seed"])
+        patience = int(train_conf.get("early_stop_patience_in_epochs"))
+        best_val_loss = float("inf")
+        patience_counter = 0
 
         training_losses = []
         validation_losses = []
         scheduler = get_constant_schedule_with_warmup(optimizer, num_warmup_steps=int(train_conf["warmup_steps"]))
 
-        train_batches = dataset["train"].batch(batch_size=int(train_conf["batch_size"]))
-        val_batches = dataset["validation"].batch(batch_size=int(train_conf["batch_size"]))
+        # Prepare dataset
+        dataset = dataset.shuffle(seed=shuffle_seed)
+
+        tokenized_train_dataset = self.__tokenize_dataset(dataset["train"], tokenizer, max_length=max_length)
+        tokenized_validation_dataset = self.__tokenize_dataset(dataset["validation"], tokenizer, max_length=max_length)
+
+        data_collator = DataCollatorForSeq2Seq(tokenizer, max_length=batch_size)
+
+        train_dataloader = DataLoader(tokenized_train_dataset, batch_size=batch_size, collate_fn=data_collator)
+        validation_dataloader = DataLoader(tokenized_validation_dataset, batch_size=batch_size, collate_fn=data_collator)
 
         self.__logger.debug("Starting the training process")
         model.train()
-        x, y, loss = None, None, None
+        loss = None
         self.__cleanup()
 
-        num_epochs = int(train_conf["num_epochs"])
-        batch_num = len(train_batches)
-        patience = int(train_conf.get("early_stop_patience_in_epochs"))
-        best_val_loss = float("inf")
-        patience_counter = 0
+        train_batch_num = len(train_dataloader)
+        val_batch_num = len(validation_dataloader)
 
-        self.__logger.debug(f"Batch number per epoch: {batch_num}")
+        train_steps = num_epochs * train_batch_num * batch_size
 
-        progress_bar = trange(num_epochs * batch_num)
+        self.__logger.debug(f"Number of epochs: {num_epochs}")
+        self.__logger.debug(f"Batch number per epoch: {train_batch_num}")
+        self.__logger.debug(f"Batch size: {batch_size}")
+        self.__logger.debug(f"TOTAL TRAINING STEPS: {train_steps}")
+
+        progress_bar = trange(train_steps)
         for epoch in range(num_epochs):
             epoch_training_losses = []
-            for i in range(batch_num):
-                batch = train_batches[i]
-                # Swap the direction of translation for some batches
-                batch = list(batch.items())
-                random.shuffle(batch)
-
-                (lang1, xx), (lang2, yy) = batch
+            for batch in train_dataloader:
+                batch = {k: v.to(model.device) for k, v in batch.items()}
```

Comment:
What are `k` and `v` other than a key and a value?

Reply:
This line is copied directly from HF documentation. Each row in the batch is a dict and we have to call `to(device)` to allow it to be loaded by cuda.

This was also done before but per row rather than per batch